src/prefix_argument/argument_trait.rs

```diff
@@ -136,56 +86,52 @@ impl<'a> PopArgumentHack<'a, serenity::Attachment> for &PhantomData<serenity::At
+    msg: &serenity::Message,
+) -> PopArgumentResult<'a, T>
+where
+    T: serenity::ArgumentConvert + Send,
+    T::Err: std::error::Error + Send + Sync + 'static,
+{
+    let (args, string) = pop_string(args).map_err(|_| (TooFewArguments::default().into(), None))?;
+    let object = T::convert(ctx, msg.guild_id, Some(msg.channel_id), &string)
+        .await
+        .map_err(|e| (e.into(), Some(string)))?;
 
-        impl std::error::Error for $error_type {}
-        impl std::fmt::Display for $error_type {
-            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                f.write_str(concat!(
-                    "Enter a valid ",
-                    stringify!($error_type),
-                    " ID or a mention."
-                ))
-            }
-        }
+    Ok((args.trim_start(), attachment_index, object))
+}
 
+/// Implements PopArgument for many types via `[pop_from_via_argumentconvert`].
+macro_rules! impl_popargument_via_argumentconvert {
+    ($($type:ty),*) => {$(
         #[async_trait::async_trait]
-        impl<'a> PopArgumentHack<'a, $type> for &PhantomData<$type> {
+        impl<'a> PopArgument<'a> for $type {
             async fn pop_from(
-                self,
                 args: &'a str,
                 attachment_index: usize,
                 ctx: &serenity::Context,
                 msg: &serenity::Message,
-            ) -> Result<
-                (&'a str, usize, $type),
-                (Box<dyn std::error::Error + Send + Sync>, Option<String>),
-            > {
-                let (args, string) =
-                    pop_string(args).map_err(|_| (TooFewArguments::default().into(), None))?;
-
-                if let Some(parsed_id) = string
-                    .parse()
-                    .ok()
-                    .or_else(|| serenity::utils::$parse_fn(&string))
-                {
-                    Ok((args.trim_start(), attachment_index, parsed_id))
-                } else {
-                    Err(($error_type::default().into(), Some(string)))
-                }
+            ) -> PopArgumentResult<'a, Self> {
+                pop_from_via_argumentconvert(args, attachment_index, ctx, msg).await
             }
         }
-    };
+    )*};
 }
 
-snowflake_pop_argument!(serenity::UserId, parse_user_mention, InvalidUserId);
-snowflake_pop_argument!(serenity::ChannelId, parse_channel_mention, InvalidChannelId);
-snowflake_pop_argument!(serenity::RoleId, parse_role_mention, InvalidRoleId);
+#[rustfmt::skip]
+impl_popargument_via_argumentconvert!(
+    f32, f64,
+    u8, u16, u32, u64,
+    i8, i16, i32, i64,
+    serenity::UserId, serenity::User, serenity::Member,
+    serenity::MessageId, serenity::Message,
+    serenity::ChannelId, serenity::Channel, serenity::GuildChannel,
+    serenity::EmojiId, serenity::Emoji,
+    serenity::RoleId, serenity::Role
+);
+#[cfg(feature = "cache")]
+impl_popargument_via_argumentconvert!(serenity::GuildId, serenity::Guild);
```

Comment:
I'm guessing these extra implementing types just come out of the aforementioned blanket impls of `T: ArgumentConvert` and subsequently `T: FromStr`.

There's a behavioral change here in that we're no longer parsing mentions of users, channels, or roles. Rather we're just parsing the raw snowflakes. Since `serenity::Mention` has a `FromStr` implementation that works for all three types, do you think we should add that to the list?

Reply:
Doesn't this implementation remove the ping support for the Id types which was done via snowflake_pop_argument?

Reply:
Which implementation are you referring to, do you mean `impl FromStr for Mention`?